File: backend/app/api/documents.py
# backend/app/api/documents.py
from typing import Literal

from fastapi import (
    APIRouter,
    Depends,
    Form,
    HTTPException,
    Response,
    UploadFile,
)
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import get_current_user
from app.core.perms import get_kb_perm, has_perm
from app.db.session import get_db
from app.models import Chunk, Document, KnowledgeBase, User
from app.schemas.document import DocumentOut
from app.services import doc_ops

router = APIRouter(tags=["documents"])


async def _require_kb_editor(db: AsyncSession, current: User,
                             kb_id: int) -> None:
    kb = await doc_ops.visible_kb_or_404(db, current, kb_id)
    if not has_perm(await get_kb_perm(db, current, kb), "editor"):
        raise HTTPException(status_code=403,
                            detail="editor permission required")


@router.post("/kbs/{kb_id}/documents", response_model=DocumentOut,
             status_code=201)
async def upload_document(
    kb_id: int,
    file: UploadFile,
    ocr: Literal["auto", "force", "off"] = Form("auto"),
    current: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await _require_kb_editor(db, current, kb_id)
    kb = await db.get(KnowledgeBase, kb_id)
    payload = await file.read()
    return await doc_ops.save_upload(
        db, kb, filename=file.filename, payload=payload,
        mime=file.content_type, ocr_mode=ocr,
        username=current.username, action="doc_upload",
    )


@router.get("/kbs/{kb_id}/documents", response_model=list[DocumentOut])
async def list_documents(
    kb_id: int,
    current: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await doc_ops.visible_kb_or_404(db, current, kb_id)
    result = await db.execute(
        select(Document).where(Document.kb_id == kb_id)
        .order_by(Document.id.desc())
    )
    return list(result.scalars().all())


@router.get("/documents/{doc_id}", response_model=DocumentOut)
async def get_document(
    doc_id: int,
    current: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await doc_ops.visible_doc_or_404(db, current, doc_id)


@router.get("/documents/{doc_id}/chunks")
async def list_chunks(
    doc_id: int,
    page: int = 1,
    page_size: int = 20,
    current: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    await doc_ops.visible_doc_or_404(db, current, doc_id)
    page = max(page, 1)
    page_size = min(max(page_size, 1), 100)
    total = (await db.execute(
        select(Chunk.id).where(Chunk.document_id == doc_id)
    )).scalars().all()
    rows = (await db.execute(
        select(Chunk).where(Chunk.document_id == doc_id)
        .order_by(Chunk.chunk_index)
        .offset((page - 1) * page_size).limit(page_size)
    )).scalars().all()
    return {
        "total": len(total),
        "items": [
            {"id": c.id, "chunk_index": c.chunk_index, "page_no": c.page_no,
             "char_len": c.char_len, "content_preview": c.content[:200]}
            for c in rows
        ],
    }


@router.post("/documents/{doc_id}/reprocess", response_model=DocumentOut)
async def reprocess_document(
    doc_id: int,
    current: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    doc = await doc_ops.visible_doc_or_404(db, current, doc_id)
    kb = await db.get(KnowledgeBase, doc.kb_id)
    if not has_perm(await get_kb_perm(db, current, kb), "editor"):
        raise HTTPException(status_code=403,
                            detail="editor permission required")
    return await doc_ops.reprocess_document(
        db, doc, username=current.username, action="doc_reprocess")


@router.delete("/documents/{doc_id}", status_code=204)
async def delete_document(
    doc_id: int,
    current: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    doc = await doc_ops.visible_doc_or_404(db, current, doc_id)
    kb = await db.get(KnowledgeBase, doc.kb_id)
    if not has_perm(await get_kb_perm(db, current, kb), "editor"):
        raise HTTPException(status_code=403,
                            detail="editor permission required")
    await doc_ops.delete_document(
        db, doc, username=current.username, action="doc_delete")
    return Response(status_code=204)

File: frontend/src/api/documents.ts
import http from './http'

/** 镜像后端 DocumentOut(app/schemas/document.py) */
export interface DocumentItem {
  id: number
  kb_id: number
  filename: string
  mime: string
  size: number
  sha256: string
  status: string
  error_msg: string | null
  page_count: number | null
  chunk_count: number
  created_at: string
}

/** 镜像 GET /documents/{id}/chunks 的条目(app/api/documents.py list_chunks) */
export interface ChunkItem {
  id: number
  chunk_index: number
  page_no: number | null
  char_len: number
  content_preview: string
}

export interface ChunksResponse {
  total: number
  items: ChunkItem[]
}

export const documentsApi = {
  async list(kbId: number): Promise<DocumentItem[]> {
    const { data } = await http.get<DocumentItem[]>(`/kbs/${kbId}/documents`)
    return data
  },

  async upload(
    kbId: number,
    file: File,
    onProgress?: (percent: number) => void,
  ): Promise<DocumentItem> {
    const form = new FormData()
    form.append('file', file)
    const { data } = await http.post<DocumentItem>(`/kbs/${kbId}/documents`, form, {
      onUploadProgress: (e) => {
        if (onProgress && e.total) {
          onProgress(Math.round((e.loaded / e.total) * 100))
        }
      },
    })
    return data
  },

  async detail(docId: number): Promise<DocumentItem> {
    const { data } = await http.get<DocumentItem>(`/documents/${docId}`)
    return data
  },

  async chunks(docId: number, page = 1, pageSize = 20): Promise<ChunksResponse> {
    const { data } = await http.get<ChunksResponse>(`/documents/${docId}/chunks`, {
      params: { page, page_size: pageSize },
    })
    return data
  },

  async reprocess(docId: number): Promise<DocumentItem> {
    const { data } = await http.post<DocumentItem>(`/documents/${docId}/reprocess`)
    return data
  },
}
